### PR TITLE
London | May 2025 | Halimatou Saddiyaa | Module-Data-Groups | Sprint 2

### PR DESCRIPTION
**Pull request by Halimatou-saddiyaa**
## Learners, PR Template

Self checklist

- [x] I have committed my files one by one, on purpose, and for a reason
- [x] I have titled my PR with REGION | COHORT_NAME | FIRST_NAME LAST_NAME | PROJ_NAME
- [x] I have tested my changes
- [x] My changes follow the [style guide](https://curriculum.codeyourfuture.io/guides/reviewing/style-guide/)
- [x] My changes meet the requirements of this task

## Changelist

I've completed the following coursework for Sprint 2:

- Debug
- implement
- Interpret

Please review my PR, thanks

## Questions

Ask any questions you have for your reviewer.

### CODE REVIEW

File: Sprint-2/implement/querystring.js [hunk 1]
```diff
@@ -1,16 +1,13 @@
...
-    return queryParams;
-  }
-  const keyValuePairs = queryString.split("&");
+  if (!queryString) return queryParams;
 
-  for (const pair of keyValuePairs) {
-    const [key, value] = pair.split("=");
-    queryParams[key] = value;
+  for (const pair of queryString.split("&")) {
+    const [key, ...rest] = pair.split("=");
```
Comment (illicitonion) [inline Sprint-2/implement/querystring.js:6]: I think you handled an edge-case here which you didn't write a test for - can you add a test for the other edge-case you fixed?

**Review by illicitonion — COMMENTED** [review]
This is looking really good! I left a few small comments to look at :)

File: Sprint-2/debug/recipe.js [hunk 2]
```diff
@@ -12,4 +15,7 @@ const recipe = {
 
 console.log(`${recipe.title} serves ${recipe.serves}
   ingredients:
-${recipe}`);
+${recipe.ingredients[0]}
```
[the remark this answers is not in the document]
Comment (Halimatou-saddiyaa) [reply Sprint-2/debug/recipe.js]: I've used a loop to make sure that the code works, whatever the number of ingredients.

**Comment by Halimatou-saddiyaa** [comment]
Thanks for your comments @illicitonion 
I've corrected my code, can you please check again?

File: Sprint-2/implement/querystring.js (hunk 1 above)
Comment (illicitonion) [reply Sprint-2/implement/querystring.js:6]: Interesting, this isn't actually the edge-case I was thinking about (though it's a good one!), I was thinking about `name=something=somethingelse` (with equals signs in the value) - can you add one for that too? :)

File: Sprint-2/debug/recipe.js [hunk 3]
```diff
@@ -4,12 +4,17 @@
...
   ingredients: ["olive oil", "tomatoes", "salt", "pepper"],
 };
 
-console.log(`${recipe.title} serves ${recipe.serves}
-  ingredients:
-${recipe}`);
+console.log(`${recipe.title} serves ${recipe.serves}`);
+console.log("  ingredients:");
+for (let ingredient of recipe.ingredients) {
+  console.log(`${ingredient}`);
```
[the remark this answers is not in the document]
Comment (Halimatou-saddiyaa) [reply Sprint-2/debug/recipe.js]: The type of each ingredient value is a string, so I probably don't need to use interpolation if I'm not adding any other values. I've changed my code accordingly.

File: Sprint-2/implement/contains.test.js [hunk 4]
```diff
@@ -17,19 +17,44 @@ as the object doesn't contains a key of 'c'
...
+test("contains on object with non-existent property returns false", () => {
+  const obj = { a: 1, b: 2 };
+  expect(contains(obj, "c")).toBe(false);
+});
+
 // Given invalid parameters like an array
 // When passed to contains
 // Then it should return false or throw an error
+
+test("contains on invalid parameters like an array returns false or throws an error", () => {
```
[the remark this answers is not in the document]
Comment (Halimatou-saddiyaa) [reply Sprint-2/implement/contains.test.js:57]: I can update the expectation of the test to be true now that I know that length is a built-in property of arrays, so that the test can pass.

File: Sprint-2/implement/querystring.js (hunk 1 above)
Comment (Halimatou-saddiyaa) [reply Sprint-2/implement/querystring.js:6]: I've added a test for the edge-case where there are multiple equals signs in the value.

File: Sprint-2/implement/contains.js [hunk 5]
```diff
@@ -1,3 +1,9 @@
-function contains() {}
+function contains(object, key) {
+  if (Array.isArray(object) && key === "length") {
```
[the remark this answers is not in the document]
Comment (Halimatou-saddiyaa) [reply Sprint-2/implement/contains.js:2]: I think Array.isArray(object) will check if the object is an array.
contains(["a", "b", "c"], "a"); should return false but using Jest, the test fails. I'm not sure why. Maybe the fact that "a" is not a key or property...
